File: binding/src/charts/libqvalueaxis.cpp
#include <QAbstractAxis>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QValueAxis>
#include <qvalueaxis.h>
#include "libqvalueaxis.h"
#include "libqvalueaxis.hxx"

QValueAxis* QValueAxis_new() {
	 return new VirtualQValueAxis();
}

QValueAxis* QValueAxis_new2(QObject* parent) {
	 return new VirtualQValueAxis(parent);
}

libqt_string QValueAxis_Tr(const char* s) {
	QString _ret = QValueAxis::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QValueAxis_Type(const QValueAxis* self) {
	return self->type();
}

void QValueAxis_SetMin(QValueAxis* self, double min) {
	self->setMin(min);
}

double QValueAxis_Min(const QValueAxis* self) {
	return self->min();
}

void QValueAxis_SetMax(QValueAxis* self, double max) {
	self->setMax(max);
}

double QValueAxis_Max(const QValueAxis* self) {
	return self->max();
}

void QValueAxis_SetRange(QValueAxis* self, double min, double max) {
	self->setRange(min, max);
}

void QValueAxis_SetTickCount(QValueAxis* self, int count) {
	self->setTickCount(count);
}

int QValueAxis_TickCount(const QValueAxis* self) {
	return self->tickCount();
}

void QValueAxis_SetMinorTickCount(QValueAxis* self, int count) {
	self->setMinorTickCount(count);
}

int QValueAxis_MinorTickCount(const QValueAxis* self) {
	return self->minorTickCount();
}

void QValueAxis_SetTickAnchor(QValueAxis* self, double anchor) {
	self->setTickAnchor(anchor);
}

double QValueAxis_TickAnchor(const QValueAxis* self) {
	return self->tickAnchor();
}

void QValueAxis_SetTickInterval(QValueAxis* self, double insterval) {
	self->setTickInterval(insterval);
}

double QValueAxis_TickInterval(const QValueAxis* self) {
	return self->tickInterval();
}

void QValueAxis_SetTickType(QValueAxis* self, int typeVal) {
	self->setTickType(static_cast<QValueAxis::TickType>(typeVal));
}

int QValueAxis_TickType(const QValueAxis* self) {
	return self->tickType();
}

void QValueAxis_SetLabelFormat(QValueAxis* self, const libqt_string format) {
	self->setLabelFormat(QString::fromUtf8(format.data, format.len));
}

libqt_string QValueAxis_LabelFormat(const QValueAxis* self) {
	QString _ret = self->labelFormat();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QValueAxis_ApplyNiceNumbers(QValueAxis* self) {
	self->applyNiceNumbers();
}

void QValueAxis_MinChanged(QValueAxis* self, double min) {
	self->minChanged(min);
}

void QValueAxis_Connect_MinChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, double) = reinterpret_cast<void (*)(QValueAxis*, double)>(slot);
    QValueAxis::connect(self, &QValueAxis::minChanged, [self, slotFunc](qreal min) {
	slotFunc(self, min);
    });
}

void QValueAxis_MaxChanged(QValueAxis* self, double max) {
	self->maxChanged(max);
}

void QValueAxis_Connect_MaxChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, double) = reinterpret_cast<void (*)(QValueAxis*, double)>(slot);
    QValueAxis::connect(self, &QValueAxis::maxChanged, [self, slotFunc](qreal max) {
	slotFunc(self, max);
    });
}

void QValueAxis_RangeChanged(QValueAxis* self, double min, double max) {
	self->rangeChanged(min, max);
}

void QValueAxis_Connect_RangeChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, double, double) = reinterpret_cast<void (*)(QValueAxis*, double, double)>(slot);
    QValueAxis::connect(self, &QValueAxis::rangeChanged, [self, slotFunc](qreal min, qreal max) {
	slotFunc(self, min, max);
    });
}

void QValueAxis_TickCountChanged(QValueAxis* self, int tickCount) {
	self->tickCountChanged(tickCount);
}

void QValueAxis_Connect_TickCountChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, int) = reinterpret_cast<void (*)(QValueAxis*, int)>(slot);
    QValueAxis::connect(self, &QValueAxis::tickCountChanged, [self, slotFunc](int tickCount) {
	slotFunc(self, tickCount);
    });
}

void QValueAxis_MinorTickCountChanged(QValueAxis* self, int tickCount) {
	self->minorTickCountChanged(tickCount);
}

void QValueAxis_Connect_MinorTickCountChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, int) = reinterpret_cast<void (*)(QValueAxis*, int)>(slot);
    QValueAxis::connect(self, &QValueAxis::minorTickCountChanged, [self, slotFunc](int tickCount) {
	slotFunc(self, tickCount);
    });
}

void QValueAxis_LabelFormatChanged(QValueAxis* self, const libqt_string format) {
	self->labelFormatChanged(QString::fromUtf8(format.data, format.len));
}

void QValueAxis_Connect_LabelFormatChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, const char*) = reinterpret_cast<void (*)(QValueAxis*, const char*)>(slot);
    QValueAxis::connect(self, &QValueAxis::labelFormatChanged, [self, slotFunc](const QString& format) {
	slotFunc(self, format);
    });
}

void QValueAxis_TickIntervalChanged(QValueAxis* self, double interval) {
	self->tickIntervalChanged(interval);
}

void QValueAxis_Connect_TickIntervalChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, double) = reinterpret_cast<void (*)(QValueAxis*, double)>(slot);
    QValueAxis::connect(self, &QValueAxis::tickIntervalChanged, [self, slotFunc](qreal interval) {
	slotFunc(self, interval);
    });
}

void QValueAxis_TickAnchorChanged(QValueAxis* self, double anchor) {
	self->tickAnchorChanged(anchor);
}

void QValueAxis_Connect_TickAnchorChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, double) = reinterpret_cast<void (*)(QValueAxis*, double)>(slot);
    QValueAxis::connect(self, &QValueAxis::tickAnchorChanged, [self, slotFunc](qreal anchor) {
	slotFunc(self, anchor);
    });
}

void QValueAxis_TickTypeChanged(QValueAxis* self, int typeVal) {
	self->tickTypeChanged(static_cast<QValueAxis::TickType>(typeVal));
}

void QValueAxis_Connect_TickTypeChanged(QValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QValueAxis*, int) = reinterpret_cast<void (*)(QValueAxis*, int)>(slot);
    QValueAxis::connect(self, &QValueAxis::tickTypeChanged, [self, slotFunc](QValueAxis::TickType typeVal) {
	slotFunc(self, typeVal);
    });
}

libqt_string QValueAxis_Tr2(const char* s, const char* c) {
	QString _ret = QValueAxis::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QValueAxis_Tr3(const char* s, const char* c, int n) {
	QString _ret = QValueAxis::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QValueAxis_QBaseType(const QValueAxis* self) {
	auto* vqvalueaxis = dynamic_cast<const VirtualQValueAxis*>(self);
	if (vqvalueaxis && vqvalueaxis->isVirtualQValueAxis) {
vqvalueaxis->setQValueAxis_Type_IsBase(true);
	return vqvalueaxis->type();
}
}

// Auxiliary method to allow providing re-implementation
void QValueAxis_OnType(const QValueAxis* self, intptr_t slot) {
	auto* vqvalueaxis = dynamic_cast<const VirtualQValueAxis*>(self);
	if (vqvalueaxis && vqvalueaxis->isVirtualQValueAxis) {
vqvalueaxis->setQValueAxis_Type_Callback(reinterpret_cast<VirtualQValueAxis::QValueAxis_Type_Callback>(slot));
}
}

void QValueAxis_Delete(QValueAxis* self) {
    delete self;
}

File: binding/include/charts/libqvalueaxis.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qvalueaxis.html
pub const qvalueaxis = struct {

    /// New constructs a new QValueAxis object.
    pub fn New() QtC.QValueAxis {
        return qtc.QValueAxis_new();
    }


    /// New2 constructs a new QValueAxis object.
    pub fn New2(parent: ?*anyopaque) QtC.QValueAxis {
        return qtc.QValueAxis_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QValueAxis_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qvalueaxis.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QValueAxis_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QValueAxis_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QValueAxis_QBaseType(@ptrCast(self));
    }

    pub fn SetMin(self: ?*anyopaque, min: f64) void {
        qtc.QValueAxis_SetMin(@ptrCast(self), min);
    }

    pub fn Min(self: ?*anyopaque, ) f64 {
        return qtc.QValueAxis_Min(@ptrCast(self));
    }

    pub fn SetMax(self: ?*anyopaque, max: f64) void {
        qtc.QValueAxis_SetMax(@ptrCast(self), max);
    }

    pub fn Max(self: ?*anyopaque, ) f64 {
        return qtc.QValueAxis_Max(@ptrCast(self));
    }

    pub fn SetRange(self: ?*anyopaque, min: f64, max: f64) void {
        qtc.QValueAxis_SetRange(@ptrCast(self), min, max);
    }

    pub fn SetTickCount(self: ?*anyopaque, count: i32) void {
        qtc.QValueAxis_SetTickCount(@ptrCast(self), count);
    }

    pub fn TickCount(self: ?*anyopaque, ) i32 {
        return qtc.QValueAxis_TickCount(@ptrCast(self));
    }

    pub fn SetMinorTickCount(self: ?*anyopaque, count: i32) void {
        qtc.QValueAxis_SetMinorTickCount(@ptrCast(self), count);
    }

    pub fn MinorTickCount(self: ?*anyopaque, ) i32 {
        return qtc.QValueAxis_MinorTickCount(@ptrCast(self));
    }

    pub fn SetTickAnchor(self: ?*anyopaque, anchor: f64) void {
        qtc.QValueAxis_SetTickAnchor(@ptrCast(self), anchor);
    }

    pub fn TickAnchor(self: ?*anyopaque, ) f64 {
        return qtc.QValueAxis_TickAnchor(@ptrCast(self));
    }

    pub fn SetTickInterval(self: ?*anyopaque, insterval: f64) void {
        qtc.QValueAxis_SetTickInterval(@ptrCast(self), insterval);
    }

    pub fn TickInterval(self: ?*anyopaque, ) f64 {
        return qtc.QValueAxis_TickInterval(@ptrCast(self));
    }

    pub fn SetTickType(self: ?*anyopaque, typeVal: i32) void {
        qtc.QValueAxis_SetTickType(@ptrCast(self), @intCast(typeVal));
    }

    pub fn TickType(self: ?*anyopaque, ) i32 {
        return qtc.QValueAxis_TickType(@ptrCast(self));
    }

    pub fn SetLabelFormat(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QValueAxis_SetLabelFormat(@ptrCast(self), format_str);
    }

    pub fn LabelFormat(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QValueAxis_LabelFormat(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qvalueaxis.LabelFormat: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ApplyNiceNumbers(self: ?*anyopaque, ) void {
        qtc.QValueAxis_ApplyNiceNumbers(@ptrCast(self));
    }

    pub fn MinChanged(self: ?*anyopaque, min: f64) void {
        qtc.QValueAxis_MinChanged(@ptrCast(self), min);
    }

    pub fn OnMinChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QValueAxis_Connect_MinChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MaxChanged(self: ?*anyopaque, max: f64) void {
        qtc.QValueAxis_MaxChanged(@ptrCast(self), max);
    }

    pub fn OnMaxChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QValueAxis_Connect_MaxChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn RangeChanged(self: ?*anyopaque, min: f64, max: f64) void {
        qtc.QValueAxis_RangeChanged(@ptrCast(self), min, max);
    }

    pub fn OnRangeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64, f64) callconv(.c) void) void {
        qtc.QValueAxis_Connect_RangeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TickCountChanged(self: ?*anyopaque, tickCount: i32) void {
        qtc.QValueAxis_TickCountChanged(@ptrCast(self), tickCount);
    }

    pub fn OnTickCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QValueAxis_Connect_TickCountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MinorTickCountChanged(self: ?*anyopaque, tickCount: i32) void {
        qtc.QValueAxis_MinorTickCountChanged(@ptrCast(self), tickCount);
    }

    pub fn OnMinorTickCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QValueAxis_Connect_MinorTickCountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LabelFormatChanged(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QValueAxis_LabelFormatChanged(@ptrCast(self), format_str);
    }

    pub fn OnLabelFormatChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QValueAxis_Connect_LabelFormatChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TickIntervalChanged(self: ?*anyopaque, interval: f64) void {
        qtc.QValueAxis_TickIntervalChanged(@ptrCast(self), interval);
    }

    pub fn OnTickIntervalChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QValueAxis_Connect_TickIntervalChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TickAnchorChanged(self: ?*anyopaque, anchor: f64) void {
        qtc.QValueAxis_TickAnchorChanged(@ptrCast(self), anchor);
    }

    pub fn OnTickAnchorChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QValueAxis_Connect_TickAnchorChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TickTypeChanged(self: ?*anyopaque, typeVal: i32) void {
        qtc.QValueAxis_TickTypeChanged(@ptrCast(self), @intCast(typeVal));
    }

    pub fn OnTickTypeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QValueAxis_Connect_TickTypeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QValueAxis_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qvalueaxis.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QValueAxis_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qvalueaxis.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QValueAxis_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/valueaxis.html#types
pub const enums = struct {
    pub const TickType = enum {
        pub const TicksDynamic: i32 = 0;
        pub const TicksFixed: i32 = 1;
    };

};
